### PR TITLE
Inrix data decoding car model working copy

### PR DESCRIPTION
**Pull request by mgsergio**
@ygorshenin

### CODE REVIEW

File: openlr/openlr_sample.cpp [hunk 1]
```diff
@@ -33,9 +33,11 @@ void ParseMWMSegments(string const & line, uint32_t const lineNumber,
     if (!strings::to_uint(segParts[2], segId))
       MYTHROW(openlr::SamplePoolLoadError, ("Can't parse MWMSegment", seg, "line:", lineNumber));
 
-    // There are two more fields in segParts: is_forward and segment length, but we ignore them.
+    bool isForward = (segParts[3] == "fwd");
```
Comment (ygorshenin) [inline openlr/openlr_sample.cpp]: const? :)

File: openlr/openlr_sample.hpp [hunk 2]
```diff
@@ -26,15 +26,16 @@ enum class ItemEvaluation
 
 struct SampleItem
 {
-  struct MWMSegemnt
+  struct MWMSegment
   {
     // TODO(mgsergio): switch to osm id.
     FeatureID m_fid;
     uint32_t m_segId;
+    bool m_isForward;
```
Comment (ygorshenin) [inline openlr/openlr_sample.hpp]: Default initialization, please.

File: qt/mainwindow.cpp [hunk 3]
```diff
@@ -172,60 +80,30 @@ class TrafficDrawerDelegate : public ITrafficDrawerDelegate
...
-
-      LOG(LINFO, ("Decoded segment", points));
-      m_drapeApi.AddLine(NextLineId(),
-                         df::DrapeApiLineData(points, dp::Color(0, 0, 255, 255))
-                         .Width(3.0f).ShowPoints(true));//.ShowId());
-    }
+    LOG(LINFO, ("Decoded segment", points));
+    m_drapeApi.AddLine(NextLineId(),
+                       df::DrapeApiLineData(points, dp::Color(0, 0, 255, 255))
+                       .Width(3.0f).ShowPoints(true));//.ShowId());
```
Comment (ygorshenin) [inline qt/mainwindow.cpp]: Add comment to the literal ```true```.

File: qt/mainwindow.cpp [hunk 4]
```diff
@@ -172,60 +80,30 @@ class TrafficDrawerDelegate : public ITrafficDrawerDelegate
...
 
   void DrawEncodedSegment(openlr::LinearSegment const & segment) override
   {
     auto const & points = segment.GetMercatorPoints();
+
+    LOG(LINFO, ("Decoded segment", points));
     m_drapeApi.AddLine(NextLineId(),
                        df::DrapeApiLineData(points, dp::Color(255, 0, 0, 255))
-                         .Width(3.0f).ShowPoints(true));//.ShowId());
+                       .Width(3.0f).ShowPoints(true));//.ShowId());
```
Comment (ygorshenin) [inline qt/mainwindow.cpp]: Do we need a comment at the end? Also, please, add a comment to the literal ```true```.

File: qt/traffic_mode.cpp [hunk 5]
```diff
@@ -28,24 +25,44 @@ DecodedSample::DecodedSample(Index const & index, openlr::SamplePool const & sam
...
   }
 }
 
-openlr::SamplePool DecodedSample::ToOpenlrSamplePool() const
+vector<m2::PointD> DecodedSample::GetPoints(size_t const index) const
 {
-  openlr::SamplePool pool;
-  for (auto const & decodedItem : m_decodedItems)
+  vector<m2::PointD> points;
+
```
Comment (ygorshenin) [inline qt/traffic_mode.cpp:35]:
```c++
auto const pushPoint = [&points](m2::PointD const & p)
{
  if (points.empty() || !AlmostEqualAbs(points.back(), p))
     points.push_back(p);
});
...
if (seg.m_isForward)
{
  pushPoint(firstP);
  pushPoint(secondP);
}
...
```

File: qt/traffic_mode.cpp [hunk 6]
```diff
@@ -28,24 +25,44 @@ DecodedSample::DecodedSample(Index const & index, openlr::SamplePool const & sam
...
+      points.push_back(secondP);
+    }
+    else
+    {
+      points.push_back(secondP);
+      points.push_back(firstP);
+    }
+  }
+
+  vector<m2::PointD> noDuplication;
```
Comment (ygorshenin) [inline qt/traffic_mode.cpp]: The name is misleading, as there actually may be duplications.

File: openlr/openlr_sample.cpp (hunk 1 above)
Comment (mgsergio) [reply openlr/openlr_sample.cpp]: Sure

File: qt/mainwindow.cpp [hunk 7]
```diff
@@ -172,60 +80,30 @@ class TrafficDrawerDelegate : public ITrafficDrawerDelegate
...
-      if (!group.m_asc)
-        points.push_back(feature.GetPoint(lastSegId + 1));
-      for (auto const & segId : group.m_segments)
-      {
-        points.push_back(feature.GetPoint(segId));
-        lastSegId = segId;
-      }
-      if (group.m_asc)
-        points.push_back(feature.GetPoint(lastSegId + 1));
+    CHECK(!sample.GetItems().empty(), ("Sample must not be empty."));
```
Comment (ygorshenin) [inline qt/mainwindow.cpp:90]: int const sampleIndex?

File: openlr/openlr_sample.cpp (hunk 1 above)
Comment (mgsergio) [reply openlr/openlr_sample.cpp]: Done

File: qt/mainwindow.cpp (hunk 4 above)
Comment (mgsergio) [reply qt/mainwindow.cpp]: Done.

File: qt/mainwindow.cpp (hunk 3 above)
Comment (mgsergio) [reply qt/mainwindow.cpp]: Done.

File: qt/traffic_mode.cpp (hunk 5 above)
Comment (mgsergio) [reply qt/traffic_mode.cpp:35]: Cool, done.

File: qt/traffic_mode.cpp (hunk 6 above)
Comment (mgsergio) [reply qt/traffic_mode.cpp]: Obsolete.

File: qt/mainwindow.cpp (hunk 7 above)
Comment (mgsergio) [reply qt/mainwindow.cpp:90]: Done.

**Comment by mgsergio** [comment]
PTAL

File: openlr/openlr_sample.hpp [hunk 8]
```diff
@@ -26,15 +26,23 @@ enum class ItemEvaluation
...
+      : m_fid(fid)
+      , m_segId(segId)
+      , m_isForward(isForward)
+    {
+    }
+
     // TODO(mgsergio): switch to osm id.
     FeatureID m_fid;
     uint32_t m_segId;
+    bool m_isForward;
```
Comment (ygorshenin) [inline openlr/openlr_sample.hpp]: bool const?

File: openlr/openlr_sample.hpp [hunk 9]
```diff
@@ -26,15 +26,23 @@ enum class ItemEvaluation
...
+    MWMSegment(FeatureID const & fid, uint32_t const segId, bool const isForward)
+      : m_fid(fid)
+      , m_segId(segId)
+      , m_isForward(isForward)
+    {
+    }
+
     // TODO(mgsergio): switch to osm id.
     FeatureID m_fid;
     uint32_t m_segId;
```
Comment (ygorshenin) [inline openlr/openlr_sample.hpp]: uint32_t const?

File: qt/traffic_mode.cpp [hunk 10]
```diff
@@ -28,24 +25,39 @@ DecodedSample::DecodedSample(Index const & index, openlr::SamplePool const & sam
...
+  auto const & item = m_decodedItems[index];
+  for (auto const & seg : item.m_segments)
   {
-    openlr::SampleItem item;
-    item.m_partnerSegmentId = decodedItem.m_partnerSegmentId;
-    item.m_evaluation = decodedItem.m_evaluation;
-    for (auto const & segment : decodedItem.m_segments)
-      item.m_segments.push_back({segment.m_fid, segment.m_segId});
-    pool.push_back(item);
+    auto const & ft = m_features.at(seg.m_fid);
```
Comment (ygorshenin) [inline qt/traffic_mode.cpp]: Do you plan to catch out of range exception? If not, just put CHECK() here.

File: openlr/openlr_sample.hpp (hunk 8 above)
Comment (mgsergio) [reply openlr/openlr_sample.hpp]: No.

File: openlr/openlr_sample.hpp (hunk 9 above)
Comment (mgsergio) [reply openlr/openlr_sample.hpp]: No.

File: qt/traffic_mode.cpp (hunk 10 above)
Comment (mgsergio) [reply qt/traffic_mode.cpp]: Done.

**Comment by mgsergio** [comment]
PTAL

File: openlr/openlr_sample.hpp (hunk 8 above)
Comment (ygorshenin) [reply openlr/openlr_sample.hpp]: Could you please be so kind to explain why?

Comment (mgsergio) [reply openlr/openlr_sample.hpp]: There is no need to do this for the specific field or a subset of fields.
MWMSegment is a value object. If you want it to be const mark the whole object as const.

Comment (ygorshenin) [reply openlr/openlr_sample.hpp]: So, why not to mark all fields as consts?

**Comment by ygorshenin** [comment]
LGTM

**Pull request merged**